replace `#[may_dangle]` with 2 attributes (#3417)
rust-lang/rfcs: text/3417-dropck-eyepatch-v3.md

- Feature Name: (`dropck_eyepatch_v3`)
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Cleanup the rules for implicit drops by splitting `#[may_dangle]` into two separate attributes:
`#[only_dropped]` and `#[fully_ignored]`. Change `PhantomData` to get completely ignored

> by dropck as its current behavior is confusing and inconsistent.

Author: these names are bikeshedable :sparkles: 

don't care too much about the specific names here

```rust
struct MyType<'s> {
    reference: &'s str,
    needs_drop: String,
}
fn can_drop_dead_reference() {
    let _x;
    {
        let temp = String::from("I am only temporary");
        _x = MyType {
            reference: &temp,
            needs_drop: String::from("I have to get dropped"),
        };
    }
    // We drop `_x` here even though `reference` is no longer live.
    //
    // This is fine as dropping a reference is a noop and does not
    // acess the pointee.
}
```
The above example will however fail if we add a manual `Drop` impl as the compiler conservatively
assumes that all generic parameters of the `Drop` impl are used:
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e604bcaecb7b2b4cf7fd0440faf165ac).

In case a manual `Drop` impl does not access a generic parameter, you can add
`#[fully_unused]` or `#[only_dropped]` to that parameter. This **unsafely** asserts
that the parameter is either completely unused when dropping your type or only
recursively dropped.

```rust
struct MyType<'s> {
    reference: &'s str,
    needs_drop: String,
}

> // The impl has to be `unsafe` as the compiler does may not check

Reviewer: ```suggestion
// The impl has to be `unsafe` as the compiler may not check
```

would never access `T` in this case.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

Whenever we use a value of a given type this type has to be **well-formed**, requiring
that all lifetimes in this type are live. An exception to this is the implicit drop when
a variable goes out of scope. While borrowck ensures that dropping the variable is safe,
this does not necessarily require all lifetimes to be live.

When implicitly dropping a variable of type `T`, liveness requirements are computed as follows:
- If `T` does not have any drop glue, do not add any requirements.
- If `T` is a trait object, `T` has to be live.
- If `T` has an explicit `Drop` impl, require all generic argument to be live, unless
    - they are marked with `#[fully_unused]`, in which case they are ignored,
    - or they are marked with `#[only_dropped]`, in which case recurse into the generic argument.
- Regardless of whether `T` implements `Drop`, recurse into all types *owned* by `T`:
    - references, raw pointers, function pointers, function items and scalars do not own
      anything. They can be trivially dropped.

>     - tuples and arrays consider their element types to be owned.

Author: for arrays this is currently also a bit inconsistent, see https://github.com/rust-lang/rust/issues/110288 and https://rust-lang.zulipchat.com/#narrow/stream/326866-t-types.2Fnominated/topic/.23110288.3A.20.60.5BT.3B.200.5D.60.20adding.20outlives.20requirements.20to.20dropck for details about this.

Shouldn't matter for this RFC apart from deciding whether to add "`[T; 0]` does not consider `T` to be owned".

- Feature Name: (`dropck_eyepatch_v3`)
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Cleanup the rules for implicit drops by adding an argument for `#[may_dangle]` on type

> parameters: `#[may_dangle(can_drop)]` and `#[may_dangle(must_not_use)]`. Change `PhantomData`

Reviewer 2: ```suggestion
parameters: `#[may_dangle(droppable)]` and `#[may_dangle(must_not_use)]`. Change `PhantomData`
```

to get completely ignored by dropck as its current behavior is confusing and inconsistent.

# Motivation
[motivation]: #motivation

The current rules around dropck and `#[may_dangle]` are confusing and have even resulted in
unsoundness [multiple](https://github.com/rust-lang/rust/issues/76367)

> [times](https://github.com/rust-lang/rust/issues/99408). Even without `#[may_dangle]`,

Reviewer 3: Re-phrasing the key point of this RFC at a slightly higher level, the argument seems to be: inferring the exact effect of `may_dangle` through field ownership is subtle and easy to get wrong, we should instead explicitly annotate what we want.

I generally agree with this direction. However, this is a departure from what seems to me as a pretty fundamental design decision in Rust: to infer all sorts of properties of a type based on "which other types it owns". We do this for dropck, but also for auto traits and variance. This decisions leads to issues such as https://github.com/rust-lang/rust/issues/99408 where changing the type of a field accidentally also changes this inferred property (that issue was `T` to `ManuallyDrop<T>` affecting dropck; I think we also had `&mut T` to `*mut T` affecting auto traits -- lucky enough erring on the safe side -- and it's easy to imagine similar situations with variance).

I am not sure how I feel about attacking this fundamental problem in a piecemeal fashion. At the very least, this RFC should state somewhere that the question of inferring may_dangle dropck from field ownership is closely related to inferring auto traits and variance. And long-term, IMO it would be very odd to have this explicit annotation for may_dangle without also having something similar for auto traits and variance. Admittedly that's a lot harder, (a) since those are stable and (b) since we only want this more explicit style when unsafe code is involved -- which for may_dangle is 100% of the cases, but not so for auto traits and variance.

Put differently: if PhantomData is not fit for the purpose of may_dangle, why is it ever fit for any purpose? Given the arguments in this RFC, it seems wrong to keep that type around (well it will stay for backwards compatibility obviously). IMO we should only accept the RFC if we generally agree that PhantomData should be deprecated and replaced by more explicit annotations.

Reviewer 4: we feel like there's no room for error with unsafe. inference isn't inherently wrong, but it should fail-safe if you forget something. that appears to hold for `Send` and `Sync` but not so much for `may_dangle`.

Reviewer 3: > that appears to hold for Send and Sync

I don't think so. If you have a `*mut T` field and later decide to make it `&mut`, you might be accidentally adding `Send`/`Sync` to your type without realizing. Similar examples exist for variance.

Reviewer 4: that's not an example, that's an argument.

we can't think of an example where changing `*mut T` to `&mut T` causes it to become unsound due to `Send`/`Sync`, but plenty of examples where it becomes unsound due to aliasing. we would love to see such an example tho.

on the other hand variance *is* a footgun, from experience. at the same time, variance is fairly unintuitive, extremely verbose, and it shows up *everywhere*, so there's a huge added value from having it inferred by default. this is in contrast to `#[may_dangle]` which (on stable) only shows up on std collections and pointer wrappers.

further, `#[may_dangle]` opts-*in* to inference (`impl Drop` is the opt-out), whereas there's no way to opt-*out* of variance.

Author: > we can't think of an example where changing `*mut T` to `&mut T` causes it to become unsound due to `Send`/`Sync`, but plenty of examples where it becomes unsound due to aliasing. we would love to see such an example tho.

There is such an example: https://github.com/rust-lang/rust/issues/41622

> Put differently: if PhantomData is not fit for the purpose of may_dangle, why is it ever fit for any purpose? Given the arguments in this RFC, it seems wrong to keep that type around (well it will stay for backwards compatibility obviously). IMO we should only accept the RFC if we generally agree that PhantomData should be deprecated and replaced by more explicit annotations.

Going to think more about that and will try to come back with an answer later.

Reviewer 3: > IMO we should only accept the RFC if we generally agree that PhantomData should be deprecated and replaced by more explicit annotations.

I worded this a bit strongly; a point can probably be made for this RFC while still keeping PhantomData for other purposes. But the RFC should then make that point, since the symptoms it describes apply more generally to field-ownership-based inference, not just to may_dangle / dropck.

Reviewer 4: > There is such an example: [rust-lang/rust#41622](https://github.com/rust-lang/rust/issues/41622)

Hm? We don't see it. That issue doesn't have any backtracking of how/when it regressed.

Reviewer 3: It wasn't a regression, but it was a case of accidental `Sync` inference: `MutexGuard` used `&mut` which propagates `Sync`; if it had used `*mut` (which arguably would be more correct as well since the lifetime of the reference is not accurate) the bug would have been avoided.

Reviewer 4: Hmm, so maybe we should have more opt-outs, but having an opt-out mixed with a partial opt-in seems like a massive footgun, especially if it's not actually checked for correctness.

But also if custom auto traits are ever to be a thing, how do they fit into all this?

Author: Alright, here are my thoughts after thinking about this during lunch:

> Put differently: if PhantomData is not fit for the purpose of may_dangle, why is it ever fit for any purpose? Given the arguments in this RFC, it seems wrong to keep that type around (well it will stay for backwards compatibility obviously). IMO we should only accept the RFC if we generally agree that PhantomData should be deprecated and replaced by more explicit annotations.

`PhantomData<T>` is not appropriate for dropck is because it always implements `Copy`. 

I think it could be appropriate if it were to only implement `Copy` if `T` is `Copy`. In this case for any `T` with drop glue we could require noop drop glue for `PhantomData` with the outlives requirements of `T`. However, such a `Copy` impl would be limiting other uses of `PhantomData` as it's quite useful to freely copy it around.

Because it implements `Copy` we're currently stuck with a type which has liveness requirements when involved in a drop, even though it doesn't get dropped itself. I personally label this behavior to clearly be a bug, hence the effort to change it via this RFC (and https://github.com/rust-lang/rust/issues/110288 for `[T; 0]`).

Reviewer 3: So you're basically saying, PhantomData for dropck is 'even weirder' than for the other purposes. I can't disagree with that, for the specific `Copy` case you mention. However that's not the argument the RFC is making, so the RFC should be amended.

> (and https://github.com/rust-lang/rust/issues/110288 for [T; 0]).

`[T; 0]` [isn't copy](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=43f5d879c6dadc661fc4fc8e8a76024a) so that argument doesn't work. I'd rather special-case array length 0 in fewer cases than more. But anyway that's not part of this RFC.

Author: updated the motivation and mentioned auto traits + variance in the future possibilities section.

In case a manual `Drop` impl does not access a generic parameter, you can add
`#[may_dangle]` to that parameter. This **unsafely** asserts
that the parameter is either completely unused when dropping your type or only
recursively dropped. For type parameters, you have to declare whether you
recursively drop instances of `T`. If so, you should use `#[may_dangle(droppable)]`.
If not, you may use `#[may_dangle(must_not_use)]`.

```rust
struct MyType<T> {
    generic: T,
    needs_drop: String,
}
// The impl has to be `unsafe` as the compiler may not check
// that `T` is actually unused.
unsafe impl<#[may_dangle(droppable)] T> Drop for MyType<T> {
    fn drop(&mut self) {
        // println!("{}", reference); // this would be unsound
        println!("{}", needs_drop);
    }
}
fn can_drop_dead_reference() {
    let _x;
    {
        let temp = String::from("I am only temporary");
        _x = MyType {
            reference: &temp,
            needs_drop: String::from("I have to get dropped"),
        };
    }
    // We drop `_x` here even though `reference` is no longer live.
    //

>     // This is accepted as `T` is marked as `#[may_dangle(can_drop)]` in the

Reviewer 2: ```suggestion
    // This is accepted as `T` is marked as `#[may_dangle(droppable)]` in the
```

`(PhantomData<PrintOnDrop<'s>>, u32)` does not. This is required for get the
behavior of `#[may_dangle(droppable)]` for parameters otherwise not owned by adding `PhantomData`
as a field. One can easily forget this, which caused the
[unsound](https://github.com/rust-lang/rust/issues/76367)
[issues](https://github.com/rust-lang/rust/issues/99408) mentioned above.

# Drawbacks
[drawbacks]: #drawbacks

This adds a small amount of implementation complexity to the compiler while not not being
fully checked and therefore requiring `unsafe`.

This RFC does not explicitly exclude stabilizing these two attributes, as they are clearer and far less
dangerous to use when compared with `#[may_dangle]`. Stabilizing these attributes will make it harder to
stabilize a more general solution like type state.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

The status quo of `#[may_dangle]` and "spooky-dropck-at-a-distance" is far from ideal and has already
resulted in unsound issues. Documenting the current behavior makes is more difficult to change later
while not officially documenting it is bound to lead to more issues and confusion going forward.

> It is therefore quite important to improve the status quo.

Reviewer 3: I don't buy this as an argument against documenting the status quo, given that this is all unstable.

Author: The issue is that it is **not**, or well `#[may_dangle]` is unstable but "spooky-dropck-at-a-distance" affects stable code:

We have the stable behavior that `(PhantomData<PrintOnDrop<'s>>, String)` requires `'s` to be live. Similar for ADTs: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f1f903239c9b472cd9982692205694cf)
```rust
use std::marker::PhantomData;
struct MyType<T, U> {
    // _marker: PhantomData<T>, // errors
    _marker: PhantomData<*const T>, // doesn't error
    other: U,
}

impl<T, U> MyType<T, U> {
    fn mk(_: T, other: U) -> Self {
        MyType {
            _marker: PhantomData,
            other,
        }
    }
}

struct PrintOnDrop<'s>(&'s str);
impl<'s> Drop for PrintOnDrop<'s> {
    fn drop(&mut self) {
        println!("dropping `{}`", self.0)
    }
}

fn main() {
    let x;
    {
        let temp = String::from("temporary");
        x = MyType::mk(PrintOnDrop(temp.as_str()), String::from("needs_drop"));
    }
}
```
While it should be fine™ to change this **stable** behavior even after documenting it as nobody should rely on it, it does make me uncomfortable.

Reviewer 3: Ah I see.
Honestly that seems like cheating to me -- the behavior hasn't been documented, but it has existed on stable for years, and it's not a bug -- this is how things were designed to behave, IIUC. Using our lack of documentation as a loophole feels a bit backhanded.

Author: > Honestly that seems like cheating to me -- the behavior hasn't been documented, but it has existed on stable for years, and it's not a bug -- this is how things were designed to behave, IIUC. Using our lack of documentation as a loophole feels a bit backhanded.

do you mean "we should keep the current behavior because it's a breaking change not to"? Or more "whether the current behavior is documented shouldn't change our decision on whether we should change it"?

> this is how things were designed to behave, IIUC

wrt to this I pretty clearly feel like this change is a [clarification of underspecified language semantics](https://rustc-dev-guide.rust-lang.org/bug-fix-procedure.html#what-qualifies-as-a-bug-fix) . I don't think we ever intended to have "does not need `Drop` but when part of something that does, add outlives requirements" as the behavior for `PhantomData`.

> "whether the current behavior is documented shouldn't change our decision on whether we should change it"

I personally disagree with this perspective if that's what you intended to say. By documenting something publicly we declare that the current behavior is intended as is and something users may rely on.

I do believe that the current state here is a bug. Dropping fields separately should be equivalent to dropping them as a pair. Once this lands I intend to add a check that `dropck_outlives_requirements.is_empty()` implies `needs_drop`. i.e. types without drop glue are trivial to drop.

Reviewer 3: > do you mean "we should keep the current behavior because it's a breaking change not to"? Or more "whether the current behavior is documented shouldn't change our decision on whether we should change it"?

The latter. I am totally fine with changing the behavior. But I don't like "we failed to document it" as justification.

- Feature Name: (`dropck_eyepatch_v3`)
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Cleanup the rules for implicit drops by adding an argument for `#[may_dangle]` on type
parameters: `#[may_dangle(droppable)]` and `#[may_dangle(must_not_use)]`. Change `PhantomData`
to get completely ignored by dropck as its current behavior is confusing and inconsistent.

# Motivation
[motivation]: #motivation

`PhantomData` is `Copy` but still adds outlives requirements when dropped as a part of

> a larger value. This behavior is inconsistent and results in "spooky-dropck-at-a-distance" 

Reviewer 5: I personally don't find this phrase ("spooky-dropck-at-a-distance") very clear and would recommend we drop it from the RFC and replace it with something more clearly meaningful. My best effort to understand what "spooky dropck at a distance means" is something like "This behavior means that the interaction of PhantomData and dropck cannot be understood in isolation, but instead has to be considered as part of the larger type that the PhantomData is embedded in". Does that sound right?

(As a stylistic side note, although I sometimes forget this myself, I generally find it's better to avoid "cutesy" negative labels for previous systems. Specifically in this case, I find I have a kneejerk reaction when I see the word "spooky" here that makes me feel defensive and want to oppose the change. I know this is silly and I am suppressing it, but it takes effort. In general, avoiding a label that feels pejorative when proposing changes will help people feel good about them.)

even without `#[may_dangle]`:

```rust

> use std::marker::PhantomData;

Reviewer 5: Stylistic note: Large blocks of code are hard for me to deal with. Can you add comments or break this into paragraphs to walk me through why each piece matters?

> struct PrintOnDrop<'a>(&'a str); // requires `'a` to be live on drop.

Reviewer 5: e.g., Consider first a type `PrintOnDrop` which holds a `&str` reference and has a `Drop` impl. This impl will access data from the reference when it is dropped. In general the compiler assumes that every drop impl will access all the data from the structs fields.

impl Drop for PrintOnDrop<'_> {
    fn drop(&mut self) {
        println!("{}", self.0)
    }
}

> fn assign<T>(_: T) -> PhantomData<T> {

Reviewer 5: explain what this is for

        println!("{}", self.needs_drop);
    }
}
fn can_drop_dead_reference() {
    let _x;
    {
        let temp = String::from("I am only temporary");
        _x = MyType {
            generic: &temp,
            needs_drop: String::from("I have to get dropped"),
        };
    }
    // We drop `_x` here even though `reference` is no longer live.
    //
    // This is accepted as `T` is marked as `#[may_dangle(droppable)]` in the
    // `Drop` impl of `MyType`.
}
```

`Drop` impls for collections tend to require `#[may_dangle(droppable)]`:

```rust
pub struct BTreeMap<K, V> {
    root: Option<Root<K, V>>,
    length: usize,
}

unsafe impl<#[may_dangle(droppable)] K, <#[may_dangle(droppable)] V> Drop for BTreeMap<K, V> {
    fn drop(&mut self) {
        // Recursively drops the key-value pairs but doesn't otherwise
        // inspect them, so we can use `#[only_dropped]` here.
        drop(unsafe {ptr::read(self) }.into_iter())
    }
}
```

A type where `#[may_dangle(must_not_use)]` would be useful is a `Weak` pointer for a variant of `Rc`
where the value is dropped when the last `Rc` goes out of scope. Dropping a `Weak` pointer

> would never access `T` in this case.

Reviewer 5: This is helpful and avoids the main problem I saw in the previous RFC -- it'd be great to expand a bit on where the two modes should be used in the stdlib to help reader understand the role of them. Also, what is an example of code that compiles because of this but wouldn't otherwise?
